### ISSUE TITLE
Lazily evaluated properties are not evaluated for `to_json` serialization

### ISSUE THREAD

**Issue by samueleaton** [crystal-lang/crystal#11918]
## Bug Report

Crystal Playground: https://play.crystal-lang.org/#/r/cxhp

```crystal
require "json"
 
class Person
  include JSON::Serializable
  property name : String
  property name_shouting : String {name.upcase + "!"}
  
  def initialize(@name)
  end
end
 
msg = Person.from_json %({"name": "Sam"})
puts msg.to_json # name_shouting is incorrectly missing
  #=> {"name":"Sam"}
msg.name_shouting
puts msg.to_json # now its visible after it was lazily assigned
  #=> {"name":"Sam","name_shouting":"SAM!"}
```
version
```
$ crystal -v
Crystal 1.3.2 (2022-01-18)
```

**Comment by Blacksmoke16** [comment]
I'm _pretty_ sure this is somewhat working as expected given `JSON::Serializable` operates on the ivar level, and it _is_ seeing this property, but it's `nil`, so it isn't (de)serialized. Which, to your point, it works fine after you call the method as by then the ivar is set.

I think you could work around this by using the `on_to_json` method provided by the `JSON::Serializable` module to serialize that property if it has not already been initialized. E.g. https://play.crystal-lang.org/#/r/cxhx.

However, quite frankly, is there really a reason this property needs to be its own ivar? Could totally just have a method that does `@name.upcase + "!"`. Then if you go that route you dont even need the conditional logic within `on_to_json` if you just always wanted to include that property.

**Comment by straight-shoota** [comment]
Yeah, I don't think this is a bug.
The current behaviour is certainly reasonable. There might be an argument to prefer different behaviour (as proposed here). But both are valid options. So this is actually a feature request to change this behaviour.
But IMO the current behaviour is better. `#to_json` is not expected to mutate the state of the object. It just serializes its current state.
If you want to serialize a different state than what you have, you should mutate it before calling `#to_json`.

I suppose the example is just a simplification and not a real use case? If you have a real use case in mind, can you share it?

**Comment by samueleaton** [comment]
Yeah i guess this is me trying to address two domains with the same class. I'm receiving upstream JSON with a `data` property that is base64 encoded. But I want to store the base64 encoding as `raw_data` and when I send that downstream to my customers, I don't want them to see the `raw_data`. 

Here I was able to hack it together.
```crystal
class Message
  include JSON::Serializable
  @[JSON::Field(key: "data", ignore_serialize: true)]
  getter raw_data : String = ""
  @[JSON::Field(ignore: true)]
  getter data : String {Base64.decode_string raw_data}

  def on_to_json(builder : JSON::Builder) : Nil
    # `data` was ignored, and `raw_data` (which is mapped to `data`) was also ignored for
    # serialization, so we add it manually
    builder.field "data", data
  end
end
```

The main problem is that I can't control the property names, I need to receive json with a `data` key, and export a json object also with a data key but that is base64 decoded. Name overlap.

**Comment by Blacksmoke16** [comment]
> is `on_to_json` running _after_ the json is built?

Yes, it's the last thing `to_json` does. Mainly intended to allow augmenting the JSON with extra info/custom logic. Another option would be to do like:

```cr
def after_initialize
  name_shouting
end
```

Which is the last thing `from_json` calls. The downside is it would make it no longer a lazy ivar :/.

> Here I was able to hack it together.

This is a pretty good solution. Don't think you'll get much better than that.

**Comment by samueleaton** [comment]
Yeah I don't always expect to serialize the objects, since its going to be exposed and I don't want to pass that small inefficiency to consumers if they just want raw data. Which is why I was making it lazy.

**Comment by straight-shoota** [comment]
This looks like a call for a custom serialize implementation tailored to each of your different use cases.

**Comment by samueleaton** [comment]
Thanks. Maybe the follow up here is to add a note to the docs on lazy properties so the user knowns that it is represented as a nilable property and my have unexpected side-effects. 

It's somewhat unintuitive that it was declared as a `: String` yet is nil and thus ignored in JSON serialization by default.

**Comment by Blacksmoke16** [comment]
> It's somewhat unintuitive that it was declared as a `: String` yet is nil and thus ignored in JSON serialization by default.

This is a side-effect of using the block `getter` macro. It makes it nilable internally because it will be `nil` until you call it, which it would then always be a `String`.
